tests/transforms/test_combine_global_phases.py

```diff
@@ -9,26 +9,136 @@
+    (transformed_qscript,), _ = combine_global_phases(qscript)
+
+    qml.assert_equal(expected_qscript, transformed_qscript)
+
+
+@pytest.mark.parametrize("phi", [-2 * np.pi, -np.pi, -1, 0, 1, np.pi, 2 * np.pi])
+def test_single_global_phase_gate(phi):
+    """Test that when the input ``QuantumScript`` has a single ``qml.GlobalPhase`` gate, the returned output has an equivalent
+    ``qml.GlobalPhase`` operation appended at the end"""
+    qscript = qml.tape.QuantumScript([qml.Hadamard(0), qml.GlobalPhase(phi, 0), qml.RX(0, 0)])
+
+    expected_qscript = qml.tape.QuantumScript([qml.Hadamard(0), qml.RX(0, 0), qml.GlobalPhase(phi)])
+    (transformed_qscript,), _ = combine_global_phases(qscript)
+
+    qml.assert_equal(expected_qscript, transformed_qscript)
+
+
+@pytest.mark.parametrize("phi1", [-2 * np.pi, -np.pi, -1, 0, 1, np.pi, 2 * np.pi])
+@pytest.mark.parametrize("phi2", [-2 * np.pi, -np.pi, -1, 0, 1, np.pi, 2 * np.pi])
+def test_multiple_global_phase_gates(phi1, phi2):
+    """Test that when the input ``QuantumScript`` has multiple ``qml.GlobalPhase`` gates, the returned output has an equivalent
+    single ``qml.GlobalPhase`` operation appended at the end with a total phase being equal to the sum of each original global phase
+    """
+    qscript = qml.tape.QuantumScript(
+        [qml.GlobalPhase(phi1, 0), qml.Hadamard(0), qml.GlobalPhase(phi2, 0), qml.RX(0, 0)]
+    )
+
+    expected_qscript = qml.tape.QuantumScript(
+        [qml.Hadamard(0), qml.RX(0, 0), qml.GlobalPhase(phi1 + phi2)]
+    )
+    (transformed_qscript,), _ = combine_global_phases(qscript)
+
+    qml.assert_equal(expected_qscript, transformed_qscript)
+
+
+@pytest.mark.parametrize("phi1", [-2 * np.pi, -np.pi, -1, 0, 1, np.pi, 2 * np.pi])
+@pytest.mark.parametrize("phi2", [-2 * np.pi, -np.pi, -1, 0, 1, np.pi, 2 * np.pi])
+def test_combine_global_phases(phi1, phi2):
+    """Test that the transform works in the autograd interface"""
+    transformed_qfunc = combine_global_phases(original_qfunc)
+
+    dev = qml.device("default.qubit", wires=3)
+    original_qnode = qml.QNode(original_qfunc, device=dev)
+    transformed_qnode = qml.QNode(transformed_qfunc, device=dev)
+
+    expected_qscript = qml.tape.make_qscript(expected_qfunc)(phi1, phi2)
+    transformed_qscript = qml.tape.make_qscript(transformed_qfunc)(phi1, phi2)
+
+    original_state = original_qnode(phi1, phi2)
+    transformed_state = transformed_qnode(phi1, phi2)
+
+    # check the equivalence between expected and transformed quantum scripts
+    qml.assert_equal(expected_qscript, transformed_qscript)
+
+    # check the equivalence between statevectors before and after the transform
+    assert np.allclose(original_state, transformed_state)
+
+
+@pytest.mark.jax
+@pytest.mark.parametrize("phi1", [-2 * np.pi, -np.pi, -1, 0, 1, np.pi, 2 * np.pi])
+@pytest.mark.parametrize("phi2", [-2 * np.pi, -np.pi, -1, 0, 1, np.pi, 2 * np.pi])
+def test_combine_global_phases_jax(phi1, phi2):
```

Comment:
The purpose of testing with different interfaces is really to confirm differentiability. You can refer to https://github.com/PennyLaneAI/pennylane/pull/6687 and look at the `TestDifferentiability` section. Basically construct a circuit with global phases, apply the transform, take the gradient (with different interfaces), and confirm that the execution does not raise an error, and that you do get 0. On a side note, for `jax` tests, we also want to test `jit` compatibility, you can see how I did it in the aforementioned PR.

Reply:
Thank you for the suggestion and for the reference to the other PR.

I think there is a quite fundamental issue here when trying to compute the gradient of a `qml.GlobalPhase` operation. As far as I know, PennyLane supports differentiation only over real-valued parameters. While it might be possible to devise a workaround to handle complex numbers, I believe doing so would not be straightforward.
In the case of a global phase rotation, the operation is defined by a complex exponential function $e^{-i\phi}$. If the parameter $\phi$ is constrained to be real, the derivative of the global phase rotation with respect to this parameter will always yield a purely imaginary value proportional to $-i e^{-i\phi}$. However, since PennyLane differentiation is designed to naturally work only with real parameters, the imaginary part of the gradient will be entirely discarded.
As a result, any attempt to compute the gradient with respect to the phase of a `qml.GlobalPhase` gate would always return zero in practice. Maybe I'm missing something, but this makes me wonder whether it even makes sense to try to compute this gradient in the first place.

Reply:
Thank you for the suggestion and for the reference to the other PR.
I would like to ask for a quick clarification before going on.

As far as I know, PennyLane supports differentiation only over real-valued parameters. In the case of a global phase rotation, since the operation is defined by a complex exponential function and the phase parameter is indeed constrained to be real, the derivative is always a purely imaginary number. This means that PennyLane differentiation entirely discard it and the output turns out to be 0 regardless of the input parameter.

Is your statement referring to this?

> confirm that the execution does not raise an error, and that you do get 0

So, should I just check it in the test?

Reply:
Resolved by commit https://github.com/PennyLaneAI/pennylane/pull/6686/commits/f0a38fd951fa9741ac717ca3f799ec854f6cbdf2.
Thank you for the suggestion and for the reference to the other PR.